Webpack complain about d.ts

<!--🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅

Oh hi there! 😄

To expedite issue processing please search open and closed issues before submitting a new one.
Existing issues often contain information about workarounds, resolution, or progress updates.

🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅-->


# 🐞 Bug report

### Command (mark with an `x`)
<!-- Can you pin-point the command or commands that are effected by this bug? -->
ng build --prod

- [ ] new
- [ x ] build
- [ ] serve
- [ ] test
- [ ] e2e
- [ ] generate
- [ ] add
- [ x ] update
- [ ] lint
- [ ] xi18n
- [ ] run
- [ ] config
- [ ] help
- [ ] version
- [ ] doc


### Is this a regression?

<!-- Did this behavior use to work in the previous version? -->
<!-- ✍️--> Yes, the previous version in which this bug was not present was: 8.2.14


### Description

I have updated my project to 9.0.0 and this happend.

I have component with type imports from `'../@types.d.ts'`, but when run build, CLI throw an error that can't resolve `'../@types'`(+ `Field 'browser' doesn't contain a valid alias configuration`). If I fixed it like `'../@types.d'` it passes, but throw another error which  `'../@types.d.ts' is missing from typescript compilation` and I should include it's path.

But things doesn't workout even though I add path to both `tsconfig.json` and `tsconfig.app.json`.
Is it because of the ivy typescript and webpack configuration?


## 🔬 Minimal Reproduction
<!--
Simple steps to reproduce this bug.

Please include: commands run (including args), packages added, related code changes.

If reproduction steps are not enough for reproduction of your issue, please create a minimal GitHub repository with the reproduction of the issue. 
A good way to make a minimal reproduction is to create a new app via `ng new repro-app` and add the minimum possible code to show the problem. 
Share the link to the repo below along with step-by-step instructions to reproduce the problem, as well as expected and actual behavior.

Issues that don't have enough info and can't be reproduced will be closed.

You can read more about issue submission guidelines here: https://github.com/angular/angular-cli/blob/master/CONTRIBUTING.md#-submitting-an-issue
-->

## 🔥 Exception or Error
<pre><code>
<!-- If the issue is accompanied by an exception or an error, please share it below: -->

Module not found: Error: Can't resolve '../@types' in '{{ project root }}\src\app\account\sign-up'
resolve '../@types' in '{{ project root }}\src\app\account\sign-up'
  using description file: {{ project root }}\package.json (relative path: ./src/app/account/sign-up)
    Field 'browser' doesn't contain a valid alias configuration
    using description file: {{ project root }}\package.json (relative path: ./src/app/account/@types)
      no extension
        Field 'browser' doesn't contain a valid alias configuration
        {{ project root }}\src\app\account\@types doesn't exist
      .ts
        Field 'browser' doesn't contain a valid alias configuration
        {{ project root }}\src\app\account\@types.ts doesn't exist
      .tsx
        Field 'browser' doesn't contain a valid alias configuration
        {{ project root }}\src\app\account\@types.tsx doesn't exist
      .mjs
        Field 'browser' doesn't contain a valid alias configuration
        {{ project root }}\src\app\account\@types.mjs doesn't exist
      .js
        Field 'browser' doesn't contain a valid alias configuration
        {{ project root }}\src\app\account\@types.js doesn't exist
      as directory
        {{ project root }}\src\app\account\@types doesn't exist
[{{ project root }}\src\app\account\@types]
[{{ project root }}\src\app\account\@types.ts]
[{{ project root }}\src\app\account\@types.tsx]
[{{ project root }}\src\app\account\@types.mjs]
[{{ project root }}\src\app\account\@types.js]
 @ ./src/app/account/sign-up/sign-up.component.ts 7:0-50
 @ ./src/app/account/account-routing.module.ts
 @ ./src/app/account/account.module.ts
 @ ./src/app/app-routing.module.ts
 @ ./src/app/app.module.ts
 @ ./src/main.ts
 @ multi ./src/main.ts

</code></pre>


## 🌍 Your Environment
<pre><code>
<!-- run `ng version` and paste output below -->

@angular-devkit/architect         0.900.1
@angular-devkit/build-angular     0.900.1
@angular-devkit/build-optimizer   0.900.1
@angular-devkit/build-webpack     0.900.1
@angular-devkit/core              9.0.1
@angular-devkit/schematics        9.0.1
@angular/cdk                      8.2.3
@angular/cli                      9.0.1
@angular/material                 8.2.3
@ngtools/webpack                  9.0.1
@schematics/angular               9.0.1
@schematics/update                0.900.1
rxjs                              6.5.4
typescript                        3.7.5
webpack                           4.41.2

</code></pre>

**Anything else relevant?**
<!-- ✍️Is this a browser specific issue? If so, please specify the browser and version. -->

<!-- ✍️Do any of these matter: operating system, IDE, package manager, HTTP server, ...? If so, please mention it below. -->

@alan-agius4 First, I have to say thank you for your affort.

Both `ng serve` and `ng build` commands either works fine for me. This become problem in production build. When I run `ng build --prod` above error occured after build all dependency modules.

```bash
$ ng build --prod

ERROR in ./src/app/account/my-account/my-account.component.ts
Module not found: Error: Can't resolve '../@types' in 'D:\Projects\eunsatio.io\front-end\blog\src\app\account\my-account'
resolve '../@types' in 'D:\Projects\eunsatio.io\front-end\blog\src\app\account\my-account'
  using description file: D:\Projects\eunsatio.io\front-end\blog\package.json (relative path: ./src/app/account/my-account)
    Field 'browser' doesn't contain a valid alias configuration
    using description file: D:\Projects\eunsatio.io\front-end\blog\package.json (relative path: ./src/app/account/@types)
      no extension
        Field 'browser' doesn't contain a valid alias configuration
        D:\Projects\eunsatio.io\front-end\blog\src\app\account\@types doesn't exist
      .ts
        Field 'browser' doesn't contain a valid alias configuration
        D:\Projects\eunsatio.io\front-end\blog\src\app\account\@types.ts doesn't exist
      .tsx
        Field 'browser' doesn't contain a valid alias configuration
        D:\Projects\eunsatio.io\front-end\blog\src\app\account\@types.tsx doesn't exist
      .mjs
        Field 'browser' doesn't contain a valid alias configuration
        D:\Projects\eunsatio.io\front-end\blog\src\app\account\@types.mjs doesn't exist
      .js
        Field 'browser' doesn't contain a valid alias configuration
        D:\Projects\eunsatio.io\front-end\blog\src\app\account\@types.js doesn't exist
      as directory
        D:\Projects\eunsatio.io\front-end\blog\src\app\account\@types doesn't exist
[D:\Projects\eunsatio.io\front-end\blog\src\app\account\@types]
[D:\Projects\eunsatio.io\front-end\blog\src\app\account\@types.ts]
[D:\Projects\eunsatio.io\front-end\blog\src\app\account\@types.tsx]
[D:\Projects\eunsatio.io\front-end\blog\src\app\account\@types.mjs]
[D:\Projects\eunsatio.io\front-end\blog\src\app\account\@types.js]
 @ ./src/app/account/my-account/my-account.component.ts 8:0-50
 @ ./src/app/account/account-routing.module.ts
 @ ./src/app/account/account.module.ts
 @ ./src/app/app-routing.module.ts
 @ ./src/app/app.module.ts
 @ ./src/main.ts
 @ multi ./src/main.ts
ERROR in ./src/app/account/sign-up/sign-up.component.ts
Module not found: Error: Can't resolve '../@types' in 'D:\Projects\eunsatio.io\front-end\blog\src\app\account\sign-up'
resolve '../@types' in 'D:\Projects\eunsatio.io\front-end\blog\src\app\account\sign-up'
  using description file: D:\Projects\eunsatio.io\front-end\blog\package.json (relative path: ./src/app/account/sign-up)
    Field 'browser' doesn't contain a valid alias configuration
    using description file: D:\Projects\eunsatio.io\front-end\blog\package.json (relative path: ./src/app/account/@types)
      no extension
        Field 'browser' doesn't contain a valid alias configuration
        D:\Projects\eunsatio.io\front-end\blog\src\app\account\@types doesn't exist
      .ts
        Field 'browser' doesn't contain a valid alias configuration
        D:\Projects\eunsatio.io\front-end\blog\src\app\account\@types.ts doesn't exist
      .tsx
        Field 'browser' doesn't contain a valid alias configuration
        D:\Projects\eunsatio.io\front-end\blog\src\app\account\@types.tsx doesn't exist
      .mjs
        Field 'browser' doesn't contain a valid alias configuration
        D:\Projects\eunsatio.io\front-end\blog\src\app\account\@types.mjs doesn't exist
      .js
        Field 'browser' doesn't contain a valid alias configuration
        D:\Projects\eunsatio.io\front-end\blog\src\app\account\@types.js doesn't exist
      as directory
        D:\Projects\eunsatio.io\front-end\blog\src\app\account\@types doesn't exist
[D:\Projects\eunsatio.io\front-end\blog\src\app\account\@types]
[D:\Projects\eunsatio.io\front-end\blog\src\app\account\@types.ts]
[D:\Projects\eunsatio.io\front-end\blog\src\app\account\@types.tsx]
[D:\Projects\eunsatio.io\front-end\blog\src\app\account\@types.mjs]
[D:\Projects\eunsatio.io\front-end\blog\src\app\account\@types.js]
 @ ./src/app/account/sign-up/sign-up.component.ts 7:0-50
 @ ./src/app/account/account-routing.module.ts
 @ ./src/app/account/account.module.ts
 @ ./src/app/app-routing.module.ts
 @ ./src/app/app.module.ts
 @ ./src/main.ts
 @ multi ./src/main.ts
```

I followed these steps and was able to reproduce:
```
git clone https://github.com/Seemspyo/Slacking-studio/
cd Slacking-studio
cd front-end/blog
npm i
ng build --optimization
```

```
kamik@RED-X1C6 MINGW64 /d/sandbox/Slacking-studio/front-end/blog (master)
$ ng build --optimization

chunk {} common.js, common.js.map (common) 2.59 kB  [rendered]
chunk {1} default~account-account-module~administrator-administrator-module~article-article-module.js, default~account-account-module~administrator-administrator-module~article-article-module.js.map (default~account-account-module~administrator-administrator-module~article-article-module) 70.6 kB  [rendered]
chunk {2} runtime.js, runtime.js.map (runtime) 2.5 kB [entry] [rendered]
chunk {3} default~account-account-module~administrator-administrator-module.js, default~account-account-module~administrator-administrator-module.js.map (default~account-account-module~administrator-administrator-module) 146 kB  [rendered]
chunk {4} account-account-module.js, account-account-module.js.map (account-account-module) 67.7 kB  [rendered]
chunk {5} administrator-administrator-module.js, administrator-administrator-module.js.map (administrator-administrator-module) 48.8 kB  [rendered]
chunk {6} article-article-module.js, article-article-module.js.map (article-article-module) 140 kB  [rendered]
chunk {7} main.js, main.js.map (main) 17.3 kB [initial] [rendered]
chunk {8} polyfills.js, polyfills.js.map (polyfills) 65 kB [initial] [rendered]
chunk {9} styles.js, styles.js.map (styles) 176 kB [initial] [rendered]
chunk {10} vendor.js, vendor.js.map (vendor) 491 kB [initial] [rendered]
Date: 2020-02-14T16:36:40.707Z - Hash: 80f8eb49dfef7eaa8167 - Time: 26623ms

ERROR in ./src/app/account/my-account/my-account.component.ts
Module not found: Error: Can't resolve '../@types' in 'D:\sandbox\Slacking-studio\front-end\blog\src\app\account\my-account'
ERROR in ./src/app/account/sign-up/sign-up.component.ts
Module not found: Error: Can't resolve '../@types' in 'D:\sandbox\Slacking-studio\front-end\blog\src\app\account\sign-up'
```

--- a/packages/ngtools/webpack/src/transformers/elide_imports.ts
+++ b/packages/ngtools/webpack/src/transformers/elide_imports.ts
@@ -39,6 +39,13 @@ export function elideImports(
       return;
     }
 
+    // Consider types for 'implements' as unused.
+    // A HeritageClause token can also be an 'AbstractKeyword'
+    // which in that case we should not elide the import.
+    if (ts.isHeritageClause(node) && node.token === ts.SyntaxKind.ImplementsKeyword) {
+      return;
+    }
+
     // Record import and skip
     if (ts.isImportDeclaration(node)) {
       imports.push(node);

--- a/packages/ngtools/webpack/src/transformers/elide_imports_spec.ts
+++ b/packages/ngtools/webpack/src/transformers/elide_imports_spec.ts
@@ -317,6 +317,27 @@ describe('@ngtools/webpack transformers', () => {
       });
     });
 
+    it(`should remove import for 'ExpressionWithTypeArguments' implements token`, () => {
+      const input = tags.stripIndent`
+        import { Bar, Buz, Unused } from './bar';
+
+        export class Foo extends Bar implements Buz { }
+
+        ${dummyNode}
+      `;
+
+      const output = tags.stripIndent`
+        import { Bar } from './bar';
+
+        export class Foo extends Bar { }
+      `;
+
+      const { program, compilerHost } = createTypescriptContext(input);
+      const result = transformTypescript(undefined, [transformer(program)], program, compilerHost);
+
+      expect(tags.oneLine`${result}`).toEqual(tags.oneLine`${output}`);
+    });
+
     describe('should not elide imports decorator type references when emitDecoratorMetadata is true', () => {
       const extraCompilerOptions: ts.CompilerOptions = {
          emitDecoratorMetadata: true,

